Multiple target item matrix

bavovanachte opened the pull request.

bavovanachte commented:
@Letme @JasperCraeghs @SteinHeselmans 

This still needs a bit of cleanup and some documentation fixes. Before I do that, I'd like some feedback on 2 points:

**The use of shlex.split instead of normal whitespace-based split here: https://github.com/melexis/sphinx-traceability-extension/pull/178/files#diff-d0297ebe2652a60819bdb6f540a89bcfR109**
  - Rationale: Column titles might have spaces in them, but we also want to provide multiple arguments for the `targettitle` configuration (which uses whitespace as a delimiter). The solution I found was to use `shlex.split`, which does splitting in a shell-like manner
  - Disadvantage: As you can see in the adapted examples, this requires adding apostrophes to titles with spaces, so could break existing documentation

**The odd behaviour of external relationships in item-matrices**
  - As illustrated in the example below, external relationships are not affected by the `target` setting (the also weren't before this MR). This can result in some bizarre matrices:

Given the following directive:

```
.. item-matrix:: Ext_toolname
    :source: r00
    :target: r00 late001
    :targettitle: "First title" "Second title"
    :nocaptions:
    :stats:
```

We get the following output, where `namespace:...` is an external relationship

![image](https://user-images.githubusercontent.com/5442584/93755419-8b46dd00-fc03-11ea-86b2-36ea17654a1d.png)

This seems very odd to me. Couple of options:

1. Keep as is.
2. Refactor the code so that external relationships are also filtered by the `target` settings
3. Remove external relationships from item-matrices altogether

bavovanachte commented:
Potential fix for https://github.com/melexis/sphinx-traceability-extension/issues/165

JasperCraeghs commented:
> The use of shlex.split instead of normal whitespace-based split here

I would use a comma as the delimiter for `:targettitle:` rather than `shlex.split` because it doesn't break compatibility. The delimiter for the `:target:` option can be a space for consistency with other directives.

> The odd behaviour of external relationships in item-matrices

I suggest to ignore external relationships when the `:type:` option is not used.

bavovanachte commented:
> I would use a comma as the delimiter for `:targettitle:` rather than `shlex.split` because it doesn't break compatibility. The delimiter for the `:target:` option can be a space for consistency with other directives

- Isn't that a bit odd from useability POV? Having all of the directives accept space-delimited lists instead of one specific setting for one specific directive?
- Also note that using comma's to split could also break existing builds that have comma's in the title name. I realize that that isn't as likely as having spaces, just saying that using commas doesn't fully _fix_ the issue of backwards compatibility
- Also note that the shlex implementation provides character escaping out of the box (`:targettitle: "First \"title\"" "Second title"` works). If we want to add commas as delimiter and still provide the ability for a user to insert commas, we have to implement character escaping for it too.

I'll happily implement the comma-based approach, I'm not arguing against it. I'd just like to weigh all the pros and cons of it. Maybe @Letme has some other arguments for or against it.
 
> I suggest to ignore external relationships when the `:type:` option is not used.

That works for me. You wouldn't expect filtering based on `target` if the external dependency `type` is passed explicitly then?

JasperCraeghs commented:
> Isn't that a bit odd from useability POV? Having all of the directives accept space-delimited lists instead of one specific setting for one specific directive?

There are several options that accept comma-separated arguments if you look at other directives. I don't think it's an issue. `shlex.split` would be completely new, however. I can see why you want to go with this approach, but I think backwards compatibility is more important than support for a comma in a title. 

> Also note that using comma's to split could also break existing builds that have comma's in the title name. I realize that that isn't as likely as having spaces, just saying that using commas doesn't fully fix the issue of backwards compatibility

If the user has only provided a single target, the plugin does not need to call `options['targettitle'].split(',')`. Backwards compatibility is not an issue if you make the plugin smart like this. After considering this argument, I'm fine with all delimiters.

>That works for me. You wouldn't expect filtering based on target if the external dependency type is passed explicitly then?

Yes, exactly.

bavovanachte commented:
> There are several options that accept comma-separated arguments if you look at other directives. I don't think it's an issue. shlex.split would be completely new, however. I can see why you want to go with this approach, but I think backwards compatibility is more important than support for a comma in a title.

Ok for me. I hadn't noticed the other directives.

> If the user has only provided a single target, the plugin does not need to call options['targettitle'].split(','). Backwards compatibility is not an issue if you make the plugin smart like this. After considering this argument, I'm fine with all delimiters.

Good point.

I think we're aligned. I'll wait a bit still for other feedback and then implement the changes.

Letme commented:
I would also say put a comma as separator here, where you want to have human readable text. It is easy not to have comma in a title, so I think we can let this option to be "not supported'. Maybe add a note in the documentation.

JasperCraeghs commented on mlx/directives/item_matrix_directive.py:
```diff
@@ -19,29 +19,42 @@ def perform_replacement(self, app, collection):
         relationships = self['type']
         if not relationships:
-            relationships = collection.iter_relations()
+            # if no explicit relationships were given, we consider all of them (expect for external ones)
```
```suggestion
            # if no explicit relationships were given, we consider all of them (except for external ones)
```

bavovanachte commented:
Ready for review. Checking the codeclimate issue now.

bavovanachte commented:
Did some refactoring in ae62059 to reduce nesting too deeply

JasperCraeghs reviewed (approved):
I added some commits to tweak the wording here and there.

bavovanachte commented:
Thanks! Is meeting the codecov/patch target strictly needed? I can certrainly do it, but as there are changes to all different directive classes to test, that would probably be a big effort (I don't think they're covered in test now?) so I prefer checking first  😄

Letme commented:
If the project diff increases we are fine - we just want upward curve (even if it is by percents..)
